pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java

```diff
@@ -167,6 +182,60 @@ private String getDefaultBrokerId() {
     }
   }
 
+  @Override
+  public Map<Long, String> getRunningQueries() {
+    return _queriesById.entrySet().stream().collect(Collectors.toMap(Map.Entry::getKey, e -> e.getValue()._query));
+  }
+
+  @VisibleForTesting
+  Set<ServerInstance> getRunningServers(long queryId) {
+    if (_queriesById.isEmpty()) {
+      return Collections.emptySet();
+    }
+    return _queriesById.get(queryId)._servers;
+  }
+
+  @Override
+  public boolean cancelQuery(long queryId, int timeoutMs, Executor executor, HttpConnectionManager connMgr,
+      Map<String, Integer> serverResponses) {
+    QueryRoutingTable routingTable = _queriesById.get(queryId);
+    if (routingTable == null) {
+      return false;
+    }
+    String globalId = getGlobalQueryId(queryId);
+    List<String> serverUrls = new ArrayList<>();
+    for (ServerInstance server : routingTable._servers) {
+      serverUrls.add(String.format("%s/query/%s", server.getAdminEndpoint(), globalId));
+    }
+    if (serverUrls.isEmpty()) {
+      LOGGER.debug("No servers running the query: {} right now", globalId);
+      return true;
```

Comment:
This usually means the routing table is not calculated. We should put a flag in the `QueryRoutingTable` (consider renaming it because it is not really routing table) to indicate whether the query is cancelled. We can check that before sending the query out.

Reply:
as explained in a comment below, if cancel request comes earlier even before query gets scheduled, then I'd suggest just to let the query continue and users can simply check running queries and make another cancel request. Cancel request is idempotent.

Reply:
renamed to QueryServers and added comment about its usage.